README.md
# PSEA: a phenotype-driven prioritizer for next-generation sequencing with high tolerance to phenotypic noise

## Install

1. obtain the source code or install from pip
   ``` bash
   git clone https://github.com/zhonghua-wang/psea.git
   ```
   ~~or install from pip~~ (TODO)
   ``` bash
   pip install psea
   ```
2. install requirements (python3)
   ```bash
   pip install requirements.txt
   ```
3. Download and build datasets
   ```bash
   python -m psea download-hpo
   python -m psea build # This step may take a few hours
   ```
## usage

Please refer to the notebook.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2022-06-16)


### Features

* add download and model building script ([1d3ce80](https://github.com/zhonghua-wang/psea/commit/1d3ce803b7c99d1a5b7352113a35410267e5cd70))


tests/test_hpo.py
import os
from psea.hpo import HPO
from psea.hpo_config import HPOConfig


def test_hpo_config():
    hpo_conf = HPOConfig()
    assert os.path.exists(hpo_conf.obo_path)
    assert os.path.exists(hpo_conf.phenotype_to_genes_path)
    assert os.path.exists(hpo_conf.genes_to_phenotype_path)
    assert os.path.exists(hpo_conf.annotation_path)
    assert hpo_conf.version.startswith('20')


def test_hpo():
    hpo = HPO()
    assert hpo.version.startswith('20')
    assert len(hpo.get_parent(hpo_id='HP:0000001')) == 0
    assert len(hpo.gene_hpo_sim_df) >= 10000
    assert hpo.get_hpo_sim_pkl('HP:0000360', 'HP:0000365') > 0


def test_psea():
    from psea.PSEA import PSEA
    psea = PSEA()
    _df = psea.predict_pkl(['HP:0000360', 'HP:0000365'])
    _dfv = psea.predict_pkl_verbose(['HP:0000360', 'HP:0000365'])
    assert psea.hpo.version.startswith('20')
    assert len(_df) >= 10
    assert 9132 in _df.index
    print(_df.head())
    print(_dfv.head())


psea/PSEA.py
import pandas as pd
import itertools as it
from psea import settings
from psea.hpo import HPO
import warnings


class PSEA(object):
    def __init__(self, hpo: HPO = None):
        if hpo is None:
            self.hpo = HPO()
        else:
            self.hpo = hpo
        self.param = settings.HPO_PARAMS.get(self.hpo.version) or settings.HPO_PARAMS['default']

    def size_rs_mean(self, size):
        return self.param['MEAN_PARAM'] * size

    def size_rs_std(self, size):
        return self.param['STD_PARAM_1'] * (size ** self.param['STD_PARAM_2'])

    def _zscore(self, rs, size):
        if size == 0:
            return -999
        return (rs - self.size_rs_mean(size)) / self.size_rs_std(size)

    def zscore(self, set1, set2):
        # raw-score
        rs = self.hpo.get_hpo_set_eric_similarity(set1, set2)
        size = len(set1) * len(set2)
        if size == 0:
            return -999
        return (rs - self.size_rs_mean(size)) / self.size_rs_std(size)

    def predict_generator(self, hpo_list):
        hpo_list = list(filter(lambda x: x in self.hpo.hpo_set, hpo_list))
        for gene, gene_hpo_list in self.hpo.gene_hpo_series.iteritems():
            yield gene, self.zscore(hpo_list, gene_hpo_list)

    def predict(self, hpo_list, sort=False):
        # remove duplicated hpo
        hpo_list = list(set(hpo_list))
        result = pd.DataFrame(list(self.predict_generator(hpo_list)), columns=['gene_id', 'zscore'])
        if sort:
            result.sort_values(by='zscore', ascending=False, inplace=True)
            result.reset_index(drop=True, inplace=True)
        return result

    @staticmethod
    def norm(_value):
        if _value == 0:
            print('debug')
        MAX = 100
        MIN = -4
        _value = MAX if _value >= MAX else _value
        _value = MIN if _value <= MIN else _value
        if _value >= 0:
            return _value / (2 * MAX) + 0.5
        else:
            return (_value - MIN) / (2 * (0 - MIN))

    def predict_pkl(self, hpo_list, sort=True, add_parent=False, norm_value=False) -> pd.DataFrame:
        """
        predict gene rank of specified hpo list
        :param hpo_list: hpo list
        :param sort: True to return sorted list
        :param add_parent: add parent hpo if hpo_list have empty gene list
        :param norm: normalize zscore to 0~1 to a new column score
        :return: predicted DataFrame
        """
        result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(set(hpo_list))]
        if len(result_df) == 0 and add_parent:
            hpo_list = list(
                it.chain.from_iterable(self.hpo.get_parent(x, valid_only=True) for x in hpo_list)
            )
            result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(hpo_list)]
        if len(result_df) == 0:
            warnings.warn(f'invalid hpo: {hpo_list}, zero will be return for all gene')
            result_df = result_df.transpose()
            result_df['zscore'] = 0
            result_df['rank'] = 0
            return result_df
        hpo_list = set(hpo_list)
        rs = result_df.sum()
        rs.name = 'rs'
        rs_df = pd.merge(rs, self.hpo.gene_hpo_size * len(hpo_list), left_index=True, right_index=True)
        rs_df['zscore'] = rs_df.apply(lambda x: self._zscore(x['rs'], x['size']), axis=1)
        rs_df.index.name = 'gene_id'
        if sort:
            rs_df.sort_values(by='zscore', ascending=False, inplace=True)
            rs_df['rank'] = range(1, len(rs_df) + 1)
        if norm_value:
            rs_df['score'] = rs_df['zscore'].map(self.norm)
        return rs_df

    def _hpo_zscore(self, _hpo_rs_seri, _hpo_list):
        _hpo_rs_seri[_hpo_list] = (_hpo_rs_seri[_hpo_list] - self.size_rs_mean(
            _hpo_rs_seri['size'])) / self.size_rs_std(len(_hpo_list) * _hpo_rs_seri['size'])
        return _hpo_rs_seri[_hpo_list]

    def predict_pkl_verbose(self, hpo_list, sort=True, return_valid_hpo=False, add_parent=False):
        if add_parent:
            hpo_map = dict((x, self.hpo.get_valid_hpo(x)) for x in set(hpo_list))
            hpo_list = it.chain.from_iterable(hpo_map.values())
        result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(hpo_list)]
        # if len(result_df) == 0 and add_parent:
        #     hpo_list = list(
        #         it.chain.from_iterable(self.hpo.get_parent(x, valid_only=True) for x in hpo_list)
        #     )
        #     result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(hpo_list)]
        new_hpo_list = result_df.index
        if len(result_df) == 0:
            raise ValueError(f'invalid hpo: {hpo_list}')

        rs_df = pd.merge(result_df.transpose(), self.hpo.gene_hpo_size, left_index=True, right_index=True)

        rs_df[new_hpo_list] = rs_df[new_hpo_list].sub(rs_df['size'].map(self.size_rs_mean), axis='index') \
            .div(rs_df['size'].map(lambda x: self.size_rs_std(x * len(new_hpo_list))), axis='index')
        rs_df['zscore'] = rs_df[new_hpo_list].sum(axis=1)
        rs_df.index.name = 'gene_id'
        rs_df = rs_df.merge(self.hpo.gene_df, how='left', left_index=True, right_index=True)
        if sort:
            rs_df.sort_values(by='zscore', ascending=False, inplace=True)
            rs_df['rank'] = range(1, len(rs_df) + 1)
        if return_valid_hpo:
            return list(new_hpo_list), rs_df
        return rs_df


if __name__ == '__main__':
    pass


psea/psea.py
import pandas as pd
import itertools as it
from psea import settings
from psea.hpo import HPO
import warnings


class PSEA(object):
    def __init__(self, hpo: HPO = None):
        if hpo is None:
            self.hpo = HPO()
        else:
            self.hpo = hpo
        self.param = settings.HPO_PARAMS[self.hpo.version]

    def size_rs_mean(self, size):
        return self.param['MEAN_PARAM'] * size

    def size_rs_std(self, size):
        return self.param['STD_PARAM_1'] * (size ** self.param['STD_PARAM_2'])

    def _zscore(self, rs, size):
        if size == 0:
            return -999
        return (rs - self.size_rs_mean(size)) / self.size_rs_std(size)

    def zscore(self, set1, set2):
        # raw-score
        rs = self.hpo.get_hpo_set_eric_similarity(set1, set2)
        size = len(set1) * len(set2)
        if size == 0:
            return -999
        return (rs - self.size_rs_mean(size)) / self.size_rs_std(size)

    def predict_generator(self, hpo_list):
        hpo_list = list(filter(lambda x: x in self.hpo.hpo_set, hpo_list))
        for gene, gene_hpo_list in self.hpo.gene_hpo_series.iteritems():
            yield gene, self.zscore(hpo_list, gene_hpo_list)

    def predict(self, hpo_list, sort=False):
        # remove duplicated hpo
        hpo_list = list(set(hpo_list))
        result = pd.DataFrame(list(self.predict_generator(hpo_list)), columns=['gene_id', 'zscore'])
        if sort:
            result.sort_values(by='zscore', ascending=False, inplace=True)
            result.reset_index(drop=True, inplace=True)
        return result

    @staticmethod
    def norm(_value):
        if _value == 0:
            print('debug')
        MAX = 100
        MIN = -4
        _value = MAX if _value >= MAX else _value
        _value = MIN if _value <= MIN else _value
        if _value >= 0:
            return _value / (2 * MAX) + 0.5
        else:
            return (_value - MIN) / (2 * (0 - MIN))

    def predict_pkl(self, hpo_list, sort=True, add_parent=False, norm_value=False):
        """
        predict gene rank of specified hpo list
        :param hpo_list: hpo list
        :param sort: True to return sorted list
        :param add_parent: add parent hpo if hpo_list have empty gene list
        :param norm: normalize zscore to 0~1 to a new column score
        :return:
        """
        result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(set(hpo_list))]
        if len(result_df) == 0 and add_parent:
            hpo_list = list(
                it.chain.from_iterable(self.hpo.get_parent(x, valid_only=True) for x in hpo_list)
            )
            result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(hpo_list)]
        if len(result_df) == 0:
            warnings.warn(f'invalid hpo: {hpo_list}, zero will be return for all gene')
            result_df = result_df.transpose()
            result_df['zscore'] = 0
            result_df['rank'] = 0
            return result_df
        hpo_list = set(hpo_list)
        rs = result_df.sum()
        rs.name = 'rs'
        rs_df = pd.merge(rs, self.hpo.gene_hpo_size * len(hpo_list), left_index=True, right_index=True)
        rs_df['zscore'] = rs_df.apply(lambda x: self._zscore(x['rs'], x['size']), axis=1)
        rs_df.index.name = 'gene_id'
        if sort:
            rs_df.sort_values(by='zscore', ascending=False, inplace=True)
            rs_df['rank'] = range(1, len(rs_df) + 1)
        if norm_value:
            rs_df['score'] = rs_df['zscore'].map(self.norm)
        return rs_df

    def _hpo_zscore(self, _hpo_rs_seri, _hpo_list):
        _hpo_rs_seri[_hpo_list] = (_hpo_rs_seri[_hpo_list] - self.size_rs_mean(
            _hpo_rs_seri['size'])) / self.size_rs_std(len(_hpo_list) * _hpo_rs_seri['size'])
        return _hpo_rs_seri[_hpo_list]

    def predict_pkl_verbose(self, hpo_list, sort=True, return_valid_hpo=False, add_parent=False):
        if add_parent:
            hpo_map = dict((x, self.hpo.get_valid_hpo(x)) for x in set(hpo_list))
            hpo_list = it.chain.from_iterable(hpo_map.values())
        result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(hpo_list)]
        # if len(result_df) == 0 and add_parent:
        #     hpo_list = list(
        #         it.chain.from_iterable(self.hpo.get_parent(x, valid_only=True) for x in hpo_list)
        #     )
        #     result_df = self.hpo.gene_hpo_sim_df[self.hpo.gene_hpo_sim_df.index.isin(hpo_list)]
        new_hpo_list = result_df.index
        if len(result_df) == 0:
            raise ValueError(f'invalid hpo: {hpo_list}')

        rs_df = pd.merge(result_df.transpose(), self.hpo.gene_hpo_size, left_index=True, right_index=True)

        rs_df[new_hpo_list] = rs_df[new_hpo_list].sub(rs_df['size'].map(self.size_rs_mean), axis='index')\
            .div(rs_df['size'].map(lambda x: self.size_rs_std(x * len(new_hpo_list))), axis='index')
        rs_df['zscore'] = rs_df[new_hpo_list].sum(axis=1)
        rs_df.index.name = 'gene_id'
        rs_df = rs_df.merge(self.hpo.gene_df, how='left', left_index=True, right_index=True)
        if sort:
            rs_df.sort_values(by='zscore', ascending=False, inplace=True)
            rs_df['rank'] = range(1, len(rs_df) + 1)
        if return_valid_hpo:
            return list(new_hpo_list), rs_df
        return rs_df


if __name__ == '__main__':
    pass
    # _hpo = HPO()
    # psea = PSEA(_hpo)
    # hpo_li = ['HP:0008619', 'HP:0004322']
    # print('verbose', psea.predict_pkl_verbose(hpo_li).head())
    # print('ori', psea.predict_pkl(hpo_li).head())


psea/pkl_prep.py
import itertools
import itertools as it
import pickle

import pandas as pd

from hpo import HPO
from tqdm import tqdm


def create_hpo_hpo_pkl(hpo_config: dict, use_full: bool, seri_full: bool):
    hpo = HPO()
    result_dict = {}
    for hpo1, hpo2 in tqdm(
            itertools.combinations_with_replacement(hpo.phen2gene_hpo_list, 2),
            desc='building HPO-HPO similarity file'
    ):
        similarity_func = hpo.get_eric_similarity_unweighted
        result_dict[(hpo1, hpo2)] = similarity_func(hpo1, hpo2)
    with open(hpo_config['hpo_hpo_pkl'], 'wb') as f:
        pickle.dump(result_dict, f)


def create_hpo_gene_pkl(hpo_config: dict, use_full: bool, seri_full: bool, weighted=False):
    hpo = HPO()

    def _func(hpo_id):
        _dict = {}
        for gene_id, _hpo_li in hpo.gene_hpo_series.items():
            _dict[gene_id] = hpo.get_hpo_set_eric_similarity(
                [hpo_id], _hpo_li)
        _seri = pd.Series(_dict)
        _seri.name = hpo_id
        return _seri

    def _func_weighted(hpo_id):
        _dict = {}
        for gene_id, _df in hpo.gene_phen_map_df.sort_values(['gene_id', 'freq_weight'], ascending=False).groupby(
                'gene_id'):
            _df = _df.drop_duplicates(['hpo_id'], keep='first')
            _dict[gene_id] = (
                    _df['hpo_id'].apply(lambda x: hpo.get_eric_similarity_unweighted(
                        hpo_id, x)) * _df['freq_weight']
            ).sum()
        _seri = pd.Series(_dict)
        _seri.name = hpo_id
        return _seri

    func = _func_weighted if weighted else _func

    _df = pd.Series(hpo.phen2gene_hpo_list,
                    hpo.phen2gene_hpo_list).parallel_apply(func)
    # return _df
    with open(hpo_config['gene_hpo_pkl'], 'wb') as w:
        pickle.dump(_df, w)


psea/settings.py
import os
import json

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CONF_DIR = os.path.join(BASE_DIR, 'conf')
# DATA_DIR = os.path.join(BASE_DIR, 'data')
LOCAL_DATA_DIR = os.path.join(BASE_DIR, '.data')
HPO_DIR = os.path.join(LOCAL_DATA_DIR, 'hpo')
HPO_LOCAL_CONF_FILE = os.path.join(LOCAL_DATA_DIR, 'hpo.json')

HPO_PARAMS_JSON_FILE = os.path.join(CONF_DIR, 'psea_params_full.json')
HPO_PARAMS = json.load(open(HPO_PARAMS_JSON_FILE))
#
# GENES_TO_PHENOTYPE_URL = 'http://purl.obolibrary.org/obo/hp/hpoa/genes_to_phenotype.txt'
# PHENOTYPE_TO_GENE_URL = 'http://purl.obolibrary.org/obo/hp/hpoa/phenotype_to_genes.txt'
#
# PHENOTYPE_ANNOTATION_URL = 'http://purl.obolibrary.org/obo/hp/hpoa/phenotype.hpoa'
# PHENOTYPE_ONTOLOGY_URL = 'http://purl.obolibrary.org/obo/hp.obo'

try:
    from psea.local_settings import *
except ImportError:
    pass



psea/training.py
import itertools
import pickle
import pandas as pd
from psea.hpo import HPO
from tqdm import tqdm
from loguru import logger

tqdm.pandas()


def create_hpo_hpo_pkl(hpo):
    result_dict = {}
    for hpo1, hpo2 in tqdm(itertools.combinations_with_replacement(
            hpo.phen2gene_hpo_list, 2), total=len(hpo.phen2gene_hpo_list) * (len(hpo.phen2gene_hpo_list) + 1) / 2
    ):
        similarity_func = hpo.get_eric_similarity_unweighted
        result_dict[(hpo1, hpo2)] = similarity_func(hpo1, hpo2)
    with open(hpo.hpo_conf.hpo_hpo_similarity_pkl_path, 'wb') as f:
        pickle.dump(result_dict, f)


def create_hpo_gene_pkl(hpo: HPO, nb_workers=1):
    def _func(hpo_id):
        _dict = {}
        for gene_id, _hpo_li in hpo.gene_hpo_series.items():
            _dict[gene_id] = hpo.get_hpo_set_eric_similarity(
                [hpo_id], _hpo_li)
        _seri = pd.Series(_dict)
        _seri.name = hpo_id
        return _seri

    def _func_weighted(hpo_id):
        _dict = {}
        for gene_id, _df in hpo.gene_phen_map_df.sort_values(['gene_id', 'freq_weight'], ascending=False).groupby(
                'gene_id'):
            _df = _df.drop_duplicates(['hpo_id'], keep='first')
            _dict[gene_id] = (
                    _df['hpo_id'].apply(lambda x: hpo.get_eric_similarity_unweighted(
                        hpo_id, x)) * _df['freq_weight']
            ).sum()
        _seri = pd.Series(_dict)
        _seri.name = hpo_id
        return _seri

    func = _func

    if nb_workers > 1:
        try:
            from pandarallel import pandarallel
            pandarallel.initialize(nb_workers=nb_workers, progress_bar=True)
            _df = pd.Series(hpo.phen2gene_hpo_list,
                            hpo.phen2gene_hpo_list).parallel_apply(func)
        except BaseException:
            Warning('Failed to run in parallel mode, using 1 worker')
            _df = pd.Series(hpo.phen2gene_hpo_list, hpo.phen2gene_hpo_list).progress_apply(func)
    else:
        _df = pd.Series(hpo.phen2gene_hpo_list, hpo.phen2gene_hpo_list).progress_apply(func)
    # return _df
    with open(hpo.hpo_conf.hpo_gene_similarity_pkl_name, 'wb') as w:
        pickle.dump(_df, w)


def build(nb_workers=1):
    hpo = HPO()
    logger.info(f'Building HPO-HPO file for HPO.{hpo.version}')
    create_hpo_hpo_pkl(hpo)
    logger.info(f'Building Gene-HPO file')
    create_hpo_gene_pkl(hpo, nb_workers=nb_workers)
    logger.info('Finish building HPO data file')


psea/hpo.py
import pandas as pd
import obonet
import networkx as nx
import math
import os

from psea.hpo_config import HPOConfig
import collections as co
import functools as ft
import itertools as it


class HPO:
    def __init__(
            self, use_full=True
            # self, use_full=True, version=settings.HPO_VERSION,
            # obo_path=settings.HPO_OBO_PATH, gene2phen_path=settings.HPO_GENE_TO_PHENOTYPE_PATH,
            # phen2gene_path=settings.HPO_PHENOTYPE_TO_GENE_PATH, gene_hpo_pkl=settings.HPO_GENE_SIM_PATH,
            # hpo_hpo_pkl=None, old=False,
    ):
        self.hpo_conf = HPOConfig()
        self.version = self.hpo_conf.version
        self._ancestor_dict = {}
        self.full_graph = obonet.read_obo(self.hpo_conf.obo_path)
        self.hpo_set = nx.ancestors(self.full_graph, 'HP:0000118')
        self.hpo_set.add('HP:0000118')
        self.graph = self.full_graph.subgraph(self.hpo_set)

        self.gene2phen = pd.read_csv(
            self.hpo_conf.genes_to_phenotype_path, skiprows=[0],
            names=['gene_id', 'gene_name', 'hpo_id', 'hpo_name', 'freq_raw', 'freq', 'info', 'source', 'source_id'],
            sep='\t'
        )

        self.phen2gene = pd.read_csv(
            self.hpo_conf.phenotype_to_genes_path, skiprows=[0],
            names=['hpo_id', 'hpo_name', 'gene_id', 'gene_name', 'info', 'source', 'source_id'], sep='\t'
        )
        self.phen2gene = self.phen2gene.loc[self.phen2gene['hpo_id'].isin(
            self.hpo_set)]
        self.gene2phen = self.gene2phen.loc[self.gene2phen['hpo_id'].isin(
            self.hpo_set)]
        if use_full:
            self.gene_phen_map_df = self.phen2gene
        else:
            self.gene_phen_map_df = self.gene2phen
        self.gene_set = set(self.phen2gene['gene_id'])
        self.gene_ic_dict = dict(
            [
                (k, -math.log(v / len(self.gene_phen_map_df)))
                for k, v in co.Counter(self.gene_phen_map_df['gene_id']).most_common()
            ]
        )
        self.hpo_ic_dict = dict(
            [
                (x[0], -math.log(x[1] / len(self.gene_phen_map_df)))
                for x in co.Counter(self.gene_phen_map_df['hpo_id']).most_common()
            ]
        )
        self.eric_ic_dict = dict(
            (hpo_id, -math.log(len(gene_list) / len(self.gene_set)))
            for hpo_id, gene_list in self.gene_phen_map_df.groupby('hpo_id')
        )

        self.phen2gene_hpo_list = list(set(self.phen2gene['hpo_id']))
        self.gene_hpo_series = self.gene_phen_map_df.groupby('gene_id')[
            'hpo_id'].apply(list)
        self.gene_hpo_size = self.gene_hpo_series.map(len)
        self.gene_hpo_size.name = 'size'

        self.gene_hpo_sim_df = None
        self.hpo_hpo_sim_df = None

        if os.path.exists(self.hpo_conf.hpo_hpo_similarity_pkl_path):
            self.hpo_hpo_sim_df = pd.read_pickle(self.hpo_conf.hpo_hpo_similarity_pkl_path)
        if os.path.exists(self.hpo_conf.hpo_gene_similarity_pkl_path):
            self.gene_hpo_sim_df = pd.read_pickle(self.hpo_conf.hpo_gene_similarity_pkl_path)

        # if gene_hpo_pkl is None or not os.path.exists(gene_hpo_pkl):
        #     self.gene_hpo_sim_df = None
        # else:
        #     with open(gene_hpo_pkl, 'rb') as f:
        #         self.gene_hpo_sim_df = pickle.load(f)
        # if hpo_hpo_pkl is None or not os.path.exists(hpo_hpo_pkl):
        #     self.hpo_hpo_sim_pkl = None
        # else:
        #     with open(hpo_hpo_pkl, 'rb') as f:
        #         self.hpo_hpo_sim_pkl = pickle.load(f)

        self.gene_df = self.gene2phen.groupby('gene_id')[['gene_name']].first()

    # def get_frequency(self, hpo_id, gene_id, include_child=True):
    #     if self._freq_dict is None:
    #         self._freq_dict = self.gene2phen.groupby(['hpo_id', 'gene_id'])['freq_weight'].agg(max).to_dict()
    #     weight = self._freq_dict.get((hpo_id, gene_id))
    #     if weight is not None or not include_child:
    #         return weight
    #
    #     weight_li = []
    #     for child in self.get_descendant_node_id_set(hpo_id):
    #         _weight = self._freq_dict.get((child, gene_id), 0)
    #         if _weight == HPOFrequency.max_weight():
    #             return _weight
    #         weight_li.append(_weight)
    #     return max(weight_li) or HPOFrequency.DEFAULT_WEIGHT

    def get_ancestor_node_id_set(self, hpo_id):
        """get ancestors hpo id set of specified HPO ID

        Arguments:
            hpo_id {str} -- HPO term id, eg. HP:0000118
        """
        return nx.descendants(self.graph, hpo_id)

    @property
    def ancestor_dict(self):
        if self._ancestor_dict == {}:
            for hpo in self.phen2gene_hpo_list:
                self._ancestor_dict[hpo] = self.get_ancestor_node_id_set(hpo)
        return self._ancestor_dict

    def get_ancestor_hpo_set(self, hpo_id):
        """
        get ancestor hpo id set of specified HPO ID
        :param hpo_id: HPO ID, i.e. HP:0000525
        :return: HPO ID set
        """
        comm = self.ancestor_dict.get(hpo_id)
        if comm is None:
            return {"HP:0000118"}
        comm.add("HP:0000118")
        comm.add(hpo_id)
        return comm

    def get_node(self, hpo_id: str, full=False):
        """get hpo entry by HPO ID

        Arguments:
            hpo_id {str} -- HPO term id, eg. HP:0000118
        """

        if full:
            return self.full_graph.nodes[hpo_id]
        return self.graph.nodes[hpo_id]

    def get_descendant_node_id_set(self, hpo_id):
        """get descendants node id set of specified HPO ID

        Arguments:
            hpo_id {str} -- HPO term id, eg. HP:0000118
        """
        return nx.ancestors(self.graph, hpo_id)

    def get_common_ancestors(self, *hpo_id_list):
        # TODO: get latest ancestor
        return ft.reduce(
            lambda x, y: x.intersection(y),
            [self.get_ancestor_hpo_set(hpo) for hpo in hpo_id_list]
        )

    def get_ic(self, hpo_id):
        return self.hpo_ic_dict.get(hpo_id)

    def get_eric_ic(self, hpo_id):
        return self.eric_ic_dict.get(hpo_id, 0)

    def get_ic_similarity(self, hpo_1, hpo_2):
        sim_list = [self.get_ic(x)
                    for x in self.get_common_ancestors(hpo_1, hpo_2)]
        sim_list = list(filter(lambda x: x is not None, sim_list))
        return max(sim_list) if len(sim_list) > 0 else 0

    def get_hpo_sim_pkl(self, hpo1, hpo2):
        return self.hpo_hpo_sim_df.get((hpo1, hpo2), 0) or self.hpo_hpo_sim_df.get((hpo2, hpo1), 0)

    def get_hpo_set_sim_pkl(self, hpo_set_1, hpo_set_2, cutoff=0):
        return sum(
            filter(
                lambda x: x >= cutoff, [self.get_hpo_sim_pkl(x, y) for x, y in it.product(hpo_set_1, hpo_set_2)]
            )
        )

    def get_set_ic_sim(self, set1, set2):
        """Calculate IC similarity (rawscore) of the specified two HPO sets

        Arguments:
            set1 {set} -- first HPO set
            set2 {set} -- second HPO set

        Returns:
            float -- IC similarity value
        """

        return sum([self.get_ic_similarity(x1, x2) for x1, x2 in it.product(set1, set2)])

    def get_parent(self, hpo_id, valid_only=False):
        """
        :param hpo_id: HPO ID, i.e. HP:0000365
        :param valid_only:
        :return:
        """
        if hpo_id == 'HP:0000001':
            return []
        if valid_only:
            try:
                parent_list = self.get_node(hpo_id).get('is_a')
            except KeyError:
                return []
            valid_li = []
            for parent_id in parent_list:
                if parent_id in self.eric_ic_dict.keys():
                    valid_li.append(parent_id)
            if len(valid_li) > 0:
                return valid_li
            else:
                for parent_id in parent_list:
                    valid_li.extend(self.get_parent(
                        parent_id, valid_only=valid_only))
            return valid_li
        try:
            return self.get_node(hpo_id).get('is_a')
        except KeyError:
            return []

    def get_parent_set(self, hpo_id):
        return set(self.graph.successors(hpo_id))

    def get_sibling_set(self, hpo_id):
        sibling_set = set(
            list(
                it.chain.from_iterable(
                    [self.get_children_set(x) for x in self.get_parent(hpo_id)]
                )
            )
        )
        sibling_set.remove(hpo_id)
        return sibling_set

    def get_children_set(self, hpo_id):
        return set(self.graph.predecessors(hpo_id))

    def get_eric_similarity_unweighted(self, hpo_id_1, hpo_id_2):
        return max(
            0,
            2 * max([self.get_eric_ic(x) for x in self.get_common_ancestors(hpo_id_1, hpo_id_2)]) - min(
                self.get_eric_ic(hpo_id_1), self.get_eric_ic(hpo_id_2))
        )

    def get_hpo_gene_similarity(self, hpo_id, gene_id):
        try:
            return self.gene_hpo_sim_df.loc[hpo_id][gene_id]
        except KeyError:
            return 0

    def get_hpo_set_eric_similarity(self, hpo_set_1, hpo_set_2, cutoff=0):
        return sum(
            filter(
                lambda x: x >= cutoff,
                [self.get_eric_similarity_unweighted(x, y) for x, y in it.product(hpo_set_1, hpo_set_2)]
            )
        )

    def gene_symbol_to_hpo(self, gene_symbol, full=False):
        if full:
            tmp_df = self.phen2gene[self.phen2gene['gene_name'] == gene_symbol]
        else:
            tmp_df = self.gene2phen[self.gene2phen['gene_name'] == gene_symbol]
        return tmp_df['hpo_id'].to_list()

    def hpo_to_gene_symbol(self, hpo_id, full=False):
        tmp_df = self.phen2gene[self.phen2gene['hpo_id'] == hpo_id] if full \
            else self.gene2phen[self.gene2phen['hpo_id'] == hpo_id]

        return tmp_df['gene_name'].to_list()

    def valid_hpo(self) -> pd.DataFrame:
        hpo_tmp = self.graph.copy()
        invalid_hpo = set(self.graph.nodes).difference(
            set(self.phen2gene_hpo_list))
        for hpo_el in invalid_hpo:
            hpo_tmp.remove_node(hpo_el)
        result_df = pd.DataFrame()
        for hpo_el in hpo_tmp.nodes:
            node = hpo_tmp.nodes[hpo_el]
            node['hpo_id'] = hpo_el
            result_df = result_df.append(node, ignore_index=True)
        return result_df

    def get_parents_id(self, hpo_id, full=True):
        hpo = self.get_node(hpo_id, full)
        if hpo.get('is_a') is None:
            return []
        li = []
        for parent_id_el in hpo['is_a']:
            li.append(parent_id_el)
            li.extend(self.get_parents_id(parent_id_el, full=True))
            # li.reverse()
        return li

    def ordered_hpo_list(self, full=False):
        li = []
        if full:
            graph = self.full_graph
        else:
            graph = self.graph
        for hpo_id in graph.nodes:
            li.append(hpo_id)
            li.extend(self.get_parents_id(hpo_id, full))
        li.reverse()
        return list(co.OrderedDict.fromkeys(li).keys())

    def get_valid_hpo(self, hpo_id):
        return self.get_parent(hpo_id, valid_only=True)

    def get_hpo_resnik_similarity(self, hpo1: str, hpo2: str):
        return max([self.get_eric_ic(x) for x in self.get_common_ancestors(hpo1, hpo2)])

    def get_hpo_set_resnik_similarity(self, hpo_set1: list, hpo_set2: list, cutoff=0):
        return sum(
            filter(lambda x: x >= cutoff,
                   [self.get_hpo_resnik_similarity(x, y) for x, y in it.product(hpo_set1, hpo_set2)])
        )
